### PR TITLE
Publish Start

### PR DESCRIPTION
**Pull request by omidontop**

### CODE REVIEW

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java [hunk 1]
```diff
@@ -223,7 +231,39 @@ public Response getReleaseJson(@QueryParam("item") CmsItemIdArg itemId,
...
+	@POST
+	@Path("api/start")
+	@Consumes(MediaType.APPLICATION_JSON)
+	@Produces(MediaType.APPLICATION_JSON)
+	public Response doStartApi(@QueryParam("item") CmsItemIdArg itemId, String body) throws Exception {
+		
+		logger.debug("Start publication requested with item: {} and options: '{}'", itemId, body);
+		
+		// Deserialize the body.
+		ObjectMapper objectMapper = new ObjectMapper();
```
Comment (takesson) [inline cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java:249]: Never instantiate ObjectMapper, kills performance. 

We always inject ObjectReader / ObjectWriter.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java [hunk 2]
```diff
@@ -108,6 +115,7 @@ public PublishResource(@Named("config:se.simonsoft.cms.hostname") String hostnam
 		this.publishConfiguration = publishConfiguration;
 		this.repackageService = repackageService;
 		this.statusService = statusService;
+		this.publishStartService = publishStartService;
```
Comment (takesson) [inline cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java:118]: Test is broken with compilation failure. See Jenkins.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+			throw new IllegalArgumentException("Requested profiling is not properly configured");
+		}
+
+		if (options.getExecutionid() == null) {
+			throw new RuntimeException("No 'executionid' was assigned");
+		}
+
+		// TODO: Need specific scenariotest setting 'startinput' but not 'startprofiling'.
+		// Set profilingRecipe.name to executionid if any of the start* parameters are provided (even if profilingRecipe == null).
+		if (options.getStartpathname() != null || options.getStartcustom().size() > 0) {
```
Comment (takesson) [inline cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java:138]: Also 'startprofiling'.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+	private PublishJob getPublishJob(CmsItemId itemId, PublishStartOptions options, PublishConfig config) {
+
+		CmsItem item = this.lookupReporting.getItem(itemId);
+		CmsItemPublish itemPublish = new CmsItemPublish(item);
+		TranslationLocalesMapping localesRfc = (TranslationLocalesMapping) this.publishConfiguration.getTranslationLocalesMapping(itemPublish);
+
+		PublishProfilingRecipe profilingRecipe = null;
+		// Use profilingname if set (get recipe from itemPublish)
+		if (itemPublish.hasProfiles() && options.getProfilingname() != null && options.getProfilingname().length() > 0) {
+			ObjectMapper objectMapper = new ObjectMapper();
```
Comment (takesson) [inline cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java:112]: Don't remove the injected ObjectMapper...

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import se.simonsoft.cms.item.command.CommandRuntimeException;
+import se.simonsoft.cms.publish.config.PublishConfiguration;
+import se.simonsoft.cms.publish.config.PublishExecutor;
+import se.simonsoft.cms.publish.config.command.PublishWebhookCommandHandler;
+import se.simonsoft.cms.publish.config.databinds.config.PublishConfig;
+import se.simonsoft.cms.publish.config.databinds.job.*;
+import se.simonsoft.cms.publish.config.databinds.profiling.PublishProfilingRecipe;
+import se.simonsoft.cms.publish.config.databinds.profiling.PublishProfilingSet;
+import se.simonsoft.cms.publish.config.item.CmsItemPublish;
+import se.simonsoft.cms.release.ProfilingSet;
```
Comment (takesson) [inline cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java:38]: I think we should use PublishProfilingSet.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java (hunk 4 above)
Comment (takesson) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java:112]: Most preferable is injecting the higher level ObjectReader.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java (hunk 5 above)
Comment (takesson) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java:38]: Right, we need the static method. Then it is correct.

These APIs need more cleanup but not this year.

**Review by takesson — CHANGES_REQUESTED** [review]
Good progress, see line comments. 

Will be interesting to see the test coverage in scenariotesting.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java [hunk 6]
```diff
@@ -223,7 +231,39 @@ public Response getReleaseJson(@QueryParam("item") CmsItemIdArg itemId,
...
+		PublishStartOptions options = objectMapper.readValue(body, PublishStartOptions.class);
+
+		// Generate and log a UUID, set into options.executionid regardless if set already.
+		UUID uuid = UUID.randomUUID();
+		options.setExecutionid(uuid.toString());
+		logger.debug("Generated executionid: '{}'", uuid);
+
+		LinkedHashMap<String, String> result = publishStartService.doPublishStartItem(itemId, options);
+
+		GenericEntity<LinkedHashMap<String, String>> ge = new GenericEntity<LinkedHashMap<String, String>>(result) {};
```
Comment (takesson) [inline cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java:259]: Probably don't need GenericEntity if we perform serialization here. The GenericEntity trick is needed when Wink resolves a REST writer implementation.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java (hunk 1 above)
Comment (omidontop) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java]: Oh, didn't know. Done.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java (hunk 3 above)
Comment (omidontop) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java]: Done.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java (hunk 4 above)
Comment (omidontop) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java]: Done.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java (hunk 6 above)
Comment (omidontop) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java]: Removed.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartOptions.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+@JsonIgnoreProperties(ignoreUnknown = true) // Allow future changes.
+public class PublishStartOptions {
+	
+	private String publication; // The config name
+	private String locale; // Select Release if empty or equal to release language.
+	private String profilingname; // One of profilingname/startprofiling.
+	private PublishProfilingRecipe startprofiling;
+	private LinkedHashMap<String, String> startcustom; // Override manifest.custom on a field-by-field basis. Validate max length in the service, alternatively size of the whole options JSON.
+	private String startpathname; // Validate max length in the service.
```
[the remark this answers is not in the document]
Comment (takesson) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartOptions.java:32]: Yes, 100 chars according to the ticket (1567). I think this validation should be in PublishStartService.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java [hunk 8]
```diff
@@ -223,7 +238,36 @@ public Response getReleaseJson(@QueryParam("item") CmsItemIdArg itemId,
...
+	 * @throws Exception
+	 */
+	@POST
+	@Path("api/start")
+	@Consumes(MediaType.APPLICATION_JSON)
+	@Produces(MediaType.APPLICATION_JSON)
+	public Response doStartApi(@QueryParam("item") CmsItemIdArg itemId, String body) throws Exception {
+		
+		logger.debug("Start publication requested with item: {} and options: '{}'", itemId, body);
+
```
[the remark this answers is not in the document]
Comment (takesson) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java:254]: Limit to 100kB according to the ticket.

Validate before parse.

**Review by takesson — CHANGES_REQUESTED** [review]

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (takesson) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java:254]: Actually limit to 110 kB. Then validate the startcustom map as well, comment below.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartOptions.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (takesson) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartOptions.java:32]: Also validate startcustom here by serializing startcustom and limit to 100kB.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+		this.translationTracking = translationTracking;
+		this.publishConfiguration = publishConfiguration;
+		this.publishExecutor = publishExecutor;
+		this.jobFactory = jobFactory;
+		this.profilingSetReader = reader.forType(PublishProfilingSet.class);
+	}
+
+
+	public LinkedHashMap<String, String> doPublishStartItem(CmsItemId itemId, PublishStartOptions options) {
+
```
Comment (takesson) [inline cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java:77]: Add validation of startpathname and startcustom here.

Probably need to inject ObjectWriter above.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartOptions.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (omidontop) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartOptions.java:32]: Done.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (omidontop) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishResource.java]: Done.

File: cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java (hunk 9 above)
Comment (takesson) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java:84]: Two things:
 - The placeholder is incorrect for String.format. I created a test in scenariontesting.
 - I realized that we should have a similar validation for startprofiling, go for 100 kB as well. So user can spend their 110kB on startcustom, startprofiling or a mix thereof.

Comment (omidontop) [reply cms-publish-rest/src/main/java/se/simonsoft/cms/publish/rest/PublishStartService.java:85]: Damn, you're right. Done.

**Review by takesson — APPROVED** [review]

**Pull request merged**